cmd/main.go

```diff
@@ -1,136 +1,173 @@
+	rootCmd.PersistentFlags().StringVarP(&port, "port", "p", "5683", "Port")
+	rootCmd.PersistentFlags().StringVarP(&auth, "auth", "a", "", "Auth")
+	rootCmd.PersistentFlags().IntVarP(&contentFormat, "content-format", "c", 50, "Content format")
+	rootCmd.PersistentFlags().StringArrayVarP(&options, "options", "0", []string{}, "Options")
 
-	os.Args = os.Args[2:]
-	o := flag.Bool("o", false, "Observe")
-	h := flag.String("h", "localhost", "Host")
-	p := flag.String("p", "5683", "Port")
-	// Default type is JSON.
-	cf := flag.Int("cf", 50, "Content format")
-	d := flag.String("d", "", "Message data")
-	a := flag.String("auth", "", "Auth token")
-	flag.Parse()
-
-	client, err := coap.New(*h + ":" + *p)
+	if err := rootCmd.Execute(); err != nil {
+		log.Fatalf("Error executing command: %v", err)
+	}
+}
+
+func makeRequest(code codes.Code, args []string) {
+	client, err := coap.New(host + ":" + port)
 	if err != nil {
-		log.Fatal("Error creating client: ", err)
+		log.Fatalf("Error coap creating client: %v", err)
 	}
 	var opts coapmsg.Options
-	if a != nil {
-		opts = append(opts, coapmsg.Option{ID: coapmsg.URIQuery, Value: []byte(fmt.Sprintf("auth=%s", *a))})
+	if options != nil {
+		for _, optString := range options {
+			opt := strings.Split(optString, ",")
+			if len(opt) < 2 {
+				log.Fatal("Invalid option format")
+			}
+			optId, err := strconv.ParseUint(opt[0], 10, 16)
+			if err != nil {
+				log.Fatal("Error parsing option id")
+			}
+			opts = append(opts, coapmsg.Option{ID: coapmsg.OptionID(optId), Value: []byte(opt[1])})
+		}
 	}
+	if auth != "" {
+		opts = append(opts, coapmsg.Option{ID: coapmsg.URIQuery, Value: []byte("auth=" + auth)})
+	}
+	switch {
+	case !observe:
+		switch code {
+		case codes.GET:
+			res, err := client.Send(args[0], code, coapmsg.MediaType(contentFormat), nil, opts...)
+			if err != nil {
+				log.Fatalf("Error sending message: %v", err)
+			}
+			printMsg(res)
 
-	if o == nil || (!*o) {
-		pld := strings.NewReader(*d)
-
-		res, err := client.Send(path, code, message.MediaType(*cf), pld, opts...)
+		default:
+			pld := strings.NewReader(data)
+			res, err := client.Send(args[0], code, coapmsg.MediaType(contentFormat), pld, opts...)
+			if err != nil {
+				log.Fatalf("Error sending message: %v", err)
+			}
+			printMsg(res)
+		}
+	case observe:
+		obs, err := client.Receive(args[0], opts...)
 		if err != nil {
-			log.Fatal("Error sending message: ", err)
+			log.Fatalf("Error observing resource: %v", err)
 		}
-		printMsg(res)
-		return
-	}
-	if code != codes.GET {
-		log.Fatal("Only GET requests accept observe option.")
+		errs := make(chan error, 2)
+		go func() {
+			c := make(chan os.Signal)
+			signal.Notify(c, syscall.SIGINT)
+			errs <- fmt.Errorf("%s", <-c)
+		}()
+
+		err = <-errs
+		obs.Cancel(context.Background())
+		log.Fatalf("Observation terminated: %v", err)
 	}
-	obs, err := client.Receive(path, opts...)
-	if err != nil {
-		log.Fatal("Error observing resource: ", err)
+}
```

Comment:
Having switch case of varabile `code` at first, make more readable for me.

```suggestion

	switch code {
	case codes.GET:
		switch {
		case observe:
			obs, err := client.Receive(args[0], opts...)
			if err != nil {
				log.Fatalf("Error observing resource: %v", err)
			}
			errs := make(chan error, 2)
			go func() {
				c := make(chan os.Signal)
				signal.Notify(c, syscall.SIGINT)
				errs <- fmt.Errorf("%s", <-c)
			}()

			err = <-errs
			obs.Cancel(context.Background())
			log.Fatalf("Observation terminated: %v", err)
		default:
			fmt.Println("got in default of get")
			res, err := client.Send(args[0], code, coapmsg.MediaType(contentFormat), nil, opts...)
			if err != nil {
				log.Fatalf("Error sending message: %v", err)
			}
			printMsg(res)
		}
	default:
		fmt.Println("got in default of post")
		pld := strings.NewReader(data)
		res, err := client.Send(args[0], code, coapmsg.MediaType(contentFormat), pld, opts...)
		if err != nil {
			log.Fatalf("Error sending message: %v", err)
		}
		printMsg(res)
	}

}
```